File: src/validation.py
"""
Pydantic models for MCP tool validation.
"""

import re
from typing import Any, Dict, List, Literal, Optional, Union

from pydantic import BaseModel, Field, field_validator, model_validator

# Regex patterns for validation
TAG_PATTERN = r'^[a-zA-Z0-9\-]+$'
ID_PATTERN = r'^[a-zA-Z0-9\-]+-[a-z2-9]{4}$'


def validate_tag_format(value: str) -> str:
    """Validate that a tag contains only alphanumeric characters and hyphens."""
    if not re.match(TAG_PATTERN, value):
        raise ValueError("Tag can only contain alphanumeric characters and hyphens")
    return value


def validate_id_format(value: str, id_type: str = "ID") -> str:
    """Validate that an ID follows the format <tag>-XXXX where XXXX is alphanumeric."""
    if not re.match(ID_PATTERN, value):
        raise ValueError(f"{id_type} must be in format <tag>-XXXX where XXXX is alphanumeric")
    return value


class TaskDict(BaseModel):
    """Dictionary representation of a task returned by the API."""
    id: str
    priority: int = Field(..., description="Plain integer priority (higher numbers = higher priority)")
    status: str = Field(..., description="Task status (ToDo or Done)")
    description: str


class MCPResponseSuccess(BaseModel):
    """Successful MCP response schema."""
    success: Literal[True] = True
    result: Any
    error: str = ""


class MCPResponseFailure(BaseModel):
    """Failed MCP response schema."""
    success: Literal[False] = False
    result: str = ""
    error: str


MCPResponse = Union[MCPResponseSuccess, MCPResponseFailure]


class PutTaskRequest(BaseModel):
    """Request model for creating or updating a task.

    Exactly one of ``tag`` or ``id`` must be provided:
    - ``tag`` (without ``id``): create a new task, generating an ID from the tag.
    - ``id`` (without ``tag``): update an existing task by ID.
    """
    description: str = Field(..., min_length=1, description="The description of the task")
    priority: int = Field(default=50, ge=0, le=9999, description="Priority from 0 (negligible) to 100 (urgent). Values outside this range are allowed for exceptional cases.")
    tag: Optional[str] = Field(None, min_length=1, max_length=12, description="Tag for new task (1-12 chars, alphanumeric and hyphens). Provide to create.")
    id: Optional[str] = Field(None, description="Existing task ID (format: <tag>-XXXX). Provide to update.")

    @field_validator('tag')
    @classmethod
    def validate_tag(cls, v):
        if v is not None:
            return validate_tag_format(v)
        return v

    @field_validator('id')
    @classmethod
    def validate_id(cls, v):
        if v is not None:
            return validate_id_format(v, "Task ID")
        return v

    @model_validator(mode='after')
    def exactly_one_of_tag_or_id(self):
        if self.tag and self.id:
            raise ValueError("Provide either 'tag' (to create) or 'id' (to update), not both")
        if not self.tag and not self.id:
            raise ValueError("Provide either 'tag' (to create) or 'id' (to update)")
        return self


class ListTasksRequest(BaseModel):
    """Request model for listing tasks."""
    count: int = Field(default=20, ge=1, le=1000, description="Maximum number of tasks to return")
    show_done: bool = Field(default=False, description="Whether to include completed tasks")


class MarkDoneRequest(BaseModel):
    """Request model for marking tasks as done."""
    task_ids: List[str] = Field(..., min_length=1, description="List of tag-based task IDs (format: <tag>-XXXX)")

    @field_validator('task_ids')
    @classmethod
    def validate_task_ids(cls, v):
        for tid in v:
            validate_id_format(tid, "Task ID")
        return v


class ListIdeasRequest(BaseModel):
    """Request model for listing ideas."""
    count: int = Field(default=20, ge=1, le=1000, description="Maximum number of ideas to return")


class PutIdeaRequest(BaseModel):
    """Request model for creating or updating an idea.

    Exactly one of ``tag`` or ``id`` must be provided:
    - ``tag`` (without ``id``): create a new idea, generating an ID from the tag.
    - ``id`` (without ``tag``): update an existing idea by ID.
    """
    score: int = Field(..., ge=0, le=9999, description="Score value (0-9999, higher numbers = higher relevance)")
    description: str = Field(..., min_length=1, description="Idea description")
    tag: Optional[str] = Field(None, min_length=1, max_length=12, description="Tag for new idea (1-12 chars, alphanumeric and hyphens). Provide to create.")
    id: Optional[str] = Field(None, description="Existing idea ID (format: <tag>-XXXX). Provide to update.")

    @field_validator('tag')
    @classmethod
    def validate_tag(cls, v):
        if v is not None:
            return validate_tag_format(v)
        return v

    @field_validator('id')
    @classmethod
    def validate_id(cls, v):
        if v is not None:
            return validate_id_format(v, "Idea ID")
        return v

    @model_validator(mode='after')
    def exactly_one_of_tag_or_id(self):
        if self.tag and self.id:
            raise ValueError("Provide either 'tag' (to create) or 'id' (to update), not both")
        if not self.tag and not self.id:
            raise ValueError("Provide either 'tag' (to create) or 'id' (to update)")
        return self


class MarkIdeaDoneRequest(BaseModel):
    """Request model for marking ideas as done."""
    idea_ids: List[str] = Field(..., min_length=1, description="List of tag-based idea IDs (format: <tag>-XXXX)")

    @field_validator('idea_ids')
    @classmethod
    def validate_idea_ids(cls, v):
        for iid in v:
            validate_id_format(iid, "Idea ID")
        return v

File: rust/src/validation.rs
//! Request validation for the MCP tool surface, mirroring the Python
//! Pydantic models' rules.

use crate::ids;

pub fn validate_tag(tag: &str) -> Result<(), String> {
    let len = tag.chars().count();
    if !(1..=12).contains(&len) {
        return Err("Tag must be 1-12 characters long".to_string());
    }
    if !tag.chars().all(|c| c.is_ascii_alphanumeric() || c == '-') {
        return Err("Tag can only contain alphanumeric characters and hyphens".to_string());
    }
    Ok(())
}

pub fn validate_id(id: &str, label: &str) -> Result<(), String> {
    if !ids::is_valid_id(id) {
        return Err(format!(
            "{label} must be in format <tag>-XXXX where XXXX is alphanumeric"
        ));
    }
    Ok(())
}

/// Exactly one of `tag` (create) or `id` (update) must be provided.
pub fn validate_exactly_one_of_tag_or_id(
    tag: &Option<String>,
    id: &Option<String>,
) -> Result<(), String> {
    match (tag, id) {
        (Some(_), Some(_)) => {
            Err("Provide either 'tag' (to create) or 'id' (to update), not both".to_string())
        }
        (None, None) => Err("Provide either 'tag' (to create) or 'id' (to update)".to_string()),
        _ => Ok(()),
    }
}

pub fn validate_range(value: i64, min: i64, max: i64, label: &str) -> Result<(), String> {
    if !(min..=max).contains(&value) {
        return Err(format!("{label} must be between {min} and {max}"));
    }
    Ok(())
}

pub fn validate_non_empty_description(description: &str) -> Result<(), String> {
    if description.is_empty() {
        return Err("Description must not be empty".to_string());
    }
    Ok(())
}

/// Validate a non-empty list of record IDs.
pub fn validate_id_list(id_list: &[String], label: &str) -> Result<(), String> {
    if id_list.is_empty() {
        return Err(format!("{label} list must not be empty"));
    }
    id_list.iter().try_for_each(|id| validate_id(id, label))
}

/// Validate the tag/id pair shared by put_task and put_idea.
pub fn validate_tag_or_id(
    tag: &Option<String>,
    id: &Option<String>,
    id_label: &str,
) -> Result<(), String> {
    if let Some(tag) = tag {
        validate_tag(tag)?;
    }
    if let Some(id) = id {
        validate_id(id, id_label)?;
    }
    validate_exactly_one_of_tag_or_id(tag, id)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn tag_rules() {
        assert!(validate_tag("fix-bug").is_ok());
        assert!(validate_tag("").is_err());
        assert!(validate_tag("thirteenchars").is_err());
        assert!(validate_tag("has space").is_err());
    }

    #[test]
    fn id_rules() {
        assert!(validate_id("tag-ab29", "Task ID").is_ok());
        assert!(validate_id("nodash", "Task ID").is_err());
        assert!(validate_id("tag-ABCD", "Task ID").is_err());
    }

    #[test]
    fn exactly_one_of_tag_or_id() {
        let some = |s: &str| Some(s.to_string());
        assert!(validate_exactly_one_of_tag_or_id(&some("t"), &None).is_ok());
        assert!(validate_exactly_one_of_tag_or_id(&None, &some("t-ab12")).is_ok());
        assert!(validate_exactly_one_of_tag_or_id(&some("t"), &some("t-ab12")).is_err());
        assert!(validate_exactly_one_of_tag_or_id(&None, &None).is_err());
    }

    #[test]
    fn ranges() {
        assert!(validate_range(0, 0, 9999, "priority").is_ok());
        assert!(validate_range(10000, 0, 9999, "priority").is_err());
        assert!(validate_range(0, 1, 1000, "count").is_err());
    }
}
